### ISSUE TITLE
Create separate Webpack configuration for Gutenberg-related entrypoints

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#4107]
Site Kit has recently added its first integration with the Gutenberg editor (for the Idea Hub notice) and will soon require much more substantial integrations (particularly for the SwG module). This brings a number of new challenges based on the following:
* Site Kit is bundling its own versions of `@wordpress/*` dependencies to be cross-compatible and not rely on what WordPress bundles.
* Gutenberg specifically relies strongly on WordPress core's own bundled versions of those.
* In order to integrate with Gutenberg, those parts of Site Kit will need to rely on WordPress core's bundled versions as well - both for compatibility (so that it even works), but also from a best-practices perspective (e.g. preferably we shouldn't load our own React there).

This issue is the first out of a few that tackle the related problems and set us up for those integrations. It focuses on creating a separate Webpack configuration for any Gutenberg-only JS assets from Site Kit.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* There should be a separate Webpack configuration for entry points that integrate with Gutenberg/the block editor.
    * Those scripts should be able to access common dependencies bundled by both Site Kit and WordPress core (in potentially diferent versions). Potentially this part could even be helpful for non-Gutenberg scripts, but it's not crucial there for now at least. If it's more straightforward to fix this globally though, let's do it.
    * Code that imports `@wordpress/*` should continue to rely on Site Kit's bundled version (i.e. the dependency that ends up in the Site Kit bundle).
    * Code that imports `@wordpress-core/*` should rely on WordPress core's bundled version (i.e. in production code this would be an external for `wp.*`). In the source code, `@wordpress-core/*` is essentially an alias for the respective `@wordpress/*` dependency coming from WordPress core. When running the non-production code (e.g. in tests), this should rely on those actual dependencies.
        * It would be nice to "double-bundle" them with different aliases (specifically related to #4108, where it would be nice to have e.g. Jest tests run with WP 5.0-bundled versions of those dependencies), but it's unclear whether that's possible.
    * Non of those scripts should have React injected.
    * In general, any of those scripts that only rely on `@wordpress-core/*` imports should not have any other Site Kit script loaded as a dependency.
* The `googlesitekit-idea-hub-notice` script (which right now is the only one that this applies to) should be adjusted to use the above Webpack configuration. The code of the file itself should also be updated to rely on `@wordpress-core/*` imports instead of `global.wp`.
    * The `@wordpress/i18n` import (`googlesitekit.i18n` in production) should be kept here for now - replacing that with `@wordpress-core/i18n` (`wp.i18n`) is part of a separate follow-up #4109.
* In PHP, the `googlesitekit-idea-hub-notice` asset should have its dependencies updated: From the Site Kit dependencies, only `googlesitekit-i18n` should remain, and instead of the other ones, all relevant WordPress core dependencies (e.g. `wp-data`) should be added to the list.

## Implementation Brief

In `webpack.config.js`, 

Add a new function 
`function customExternalsBundle( mode, entryPoints, externalDependencies )`

Where `mode` is a string like the following:
https://github.com/google/site-kit-wp/blob/eb6d7750c4093c715efded685db60e0fdbb66536/webpack.config.js#L455

`entryPoints` is an object of the following format:
https://github.com/google/site-kit-wp/blob/eb6d7750c4093c715efded685db60e0fdbb66536/webpack.config.js#L329-L334

`externalDependencies` is an object of the following format:
https://github.com/google/site-kit-wp/blob/eb6d7750c4093c715efded685db60e0fdbb66536/webpack.config.js#L85-L91

The implementation will mirror the existing config in `webpackConfig()`  https://github.com/google/site-kit-wp/blob/eb6d7750c4093c715efded685db60e0fdbb66536/webpack.config.js#L160-L321
With some minor tweaks to:

- Use the supplied `mode`, `entryPoints` (keyed with `entry`), and `externalDependencies` (keyed with `externals`) params 

Create a new object `gutenbergEntryPoints` which will contain the following entry (`googlesitekit-idea-hub-notice` remove it from where it currently is): 
https://github.com/google/site-kit-wp/blob/eb6d7750c4093c715efded685db60e0fdbb66536/webpack.config.js#L199-L200

Create a new object `gutenbergExternals` which will contain the following entry (`@wordpress/i18n` don't remove it from where it currently is): 
https://github.com/google/site-kit-wp/blob/eb6d7750c4093c715efded685db60e0fdbb66536/webpack.config.js#L90

```js
configs.push( { ...customExternalsBundle( mode, gutenbergEntryPoints, gutenbergExternals ) } );

```
Add to the configs with a call to the new function:
https://github.com/google/site-kit-wp/blob/eb6d7750c4093c715efded685db60e0fdbb66536/webpack.config.js#L463-L464

Finally update dependencies in `includes/Modules/Idea_Hub.php`, removing everything except `googlesitekit-i18n`, next add the following dependencies:
 - wp-data
 - wp-api-fetch
 - wp-polyfill
 - wp-url

https://github.com/google/site-kit-wp/blob/eb6d7750c4093c715efded685db60e0fdbb66536/includes/Modules/Idea_Hub.php#L728-L732

### Test Coverage

* N/A

## QA Brief

* Ensure Site Kit dashboard, admin bar, and setup still load normally.
* Create an Idea Hub draft, visit the draft in the post editor (with Gutenberg enabled), and ensure the notice still appears:

<img width="971" alt="Screenshot 2021-12-10 at 22 55 56" src="https://user-images.githubusercontent.com/90871/145651515-aef8b1ac-1f91-4f9f-b48c-fe130ebb2e3b.png">

## Changelog entry

* Use only WordPress core's bundled dependencies instead of Site Kit's when integrating with the block editor.

**Comment by felixarntz** [comment]
@aaemnnosttv @eugene-manuilov @tofumatt @ChrisAntaki I've added ACs here with some requirements. Please have a look and let me know if there's anything to add.

**Comment by tofumatt** [comment]
The ACs here seem fine to me—I don't think there's anything else to cover. 👍🏻

**Comment by tofumatt** [comment]
I think the approach here makes sense—I know @aaemnnosttv is working on making our datastore registry accessible through WordPress Data (at least when using Gutenberg), but the approach here works.

I guess one thing to check is whether we use anything in a `wp-*` packages aside from `wp-element` that isn't available in WP 5.0 and will cause errors. But that's out-of-scope for this issue, just thinking aloud really 😅

IB ✅

**Comment by aaemnnosttv** [comment]
@tofumatt as far as imports go, I think everything we use is available on the `wp` global in 5.0. There may be specific components, functionality, etc that has changed though but nothing comes to mind aside from WP data (5.0 didn't have `__experimentalResolveSelect` for example).

**Comment by kuasha420** [comment]
@aaemnnosttv @tofumatt I am seeing some inconsistencies between AC and IB here, some of them are:

* IB has no mention of import-aliasing `@wordpress-core/*`.
* IB has no mention of updating `assets/js/googlesitekit-idea-hub-notice.js`.
* It's not clear from the IB which plugins we want for the gutenberg entrypoints or do we want all of them. 

Some clarification about above point would be appreciated.

**Comment by tofumatt** [comment]
@kuasha420 Argh, that's totally correct. I think I overlooked it because I had discussed the approach we'd likely use before but I can see it's missing from the IB. My brain must've filled in the blanks, sorry!

The approach here should be that we should alias all `@wordpress-core/*` imports to use the relevant `window.wp.*` package. We'll want to alias `@wordpress-core/api-fetch` to `wp.apiFetch`, `@wordpress-core/element` to `wp.element`, etc. Fortunately we actually _used_ to do this so there's prior art in the repo for the approach: https://github.com/google/site-kit-wp/blob/9a262cd18c33995ce5ec81bc300ff055dff2a153/webpack.config.js#L32-L85

`assets/js/googlesitekit-idea-hub-notice.js` should be updated to use the new `@wordpress-core/*` imports so it eventually resolves to the built-in `wp.*` objects.

> It's not clear from the IB which plugins we want for the gutenberg entrypoints or do we want all of them.

I'm not sure what you mean here—do you mean which packages should we create aliases for? I think for now the list in the link I posted above is fine, but if any are missing feel free to add them. If it's trivial to add them all (or if it can be automated) then go for that I guess 😄

**Comment by aaemnnosttv** [comment]
> I'm not sure what you mean here—do you mean which packages should we create aliases for?

@tofumatt I think he means the `plugins` for the bundle (e.g. `ProvidePlugin`, etc). Correct @kuasha420 ?

**Comment by kuasha420** [comment]
@aaemnnosttv yep, I meant the webpack plugins in the `plugins` array. The webpack config is one busy file, so I can see why it may get lost in translation.

Other than that, everything else is cleared for the most part and rest can get sorted in CR. 

cc @tofumatt

**Comment by tofumatt** [comment]
Ah, thanks! Sorry I missed that; in terms of `plugins` to use, I think we want everything here _except_ the `ProvidePlugin`:

https://github.com/google/site-kit-wp/blob/9683b8c3687350ba6134b1a484285ceb12a236e1/webpack.config.js#L239-L278

We don't want to provide React in the bundle ourselves; users should be using the React from WordPress Element there.

The `WebpackBar` option should be edited to show a new colour (dealer's choice as to what looks good there) and a name like "Editor Entry Points" or similar.

**Comment by kuasha420** [comment]
@tofumatt I have created a PR with my current progress so far. I think we can sort out the rest in the CR (assuming I haven't done it completely wrong :grimacing: ). Would you mind reviewing it? 

Cheers.

**Comment by tofumatt** [comment]
I've amended the PR with a fix, but I've changed enough code that I probably shouldn't be the one to review it now, so un-assigning myself from reviewing this one 😅

**Comment by kuasha420** [comment]
Thanks @tofumatt. Just adding that the PR has deviated a bit from the IB (no customBundle function) to make sure everything works.

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@kuasha420 just one observation. I see a slight difference than the screenshot in the QAB. 

The WordPress top menu bar does not appear when I am in editing a post. 

This is what I see: 

![image](https://user-images.githubusercontent.com/73545194/146196282-128ffdbd-81ae-4409-bfe6-5c359b0d9523.png)

This might be more to do with my test site because even on the latest release it doesn't appear, but wanted to check.

**Comment by tofumatt** [comment]
@wpdarren Yes, that's a feature of newer Post Editor versions and not related to this change. The screenshot in the QAB is just from an older version of WordPress so the top bar is still there.

But here's my site without this code, it shows the same:

<img width="1142" alt="Screenshot 2021-12-15 at 13 53 38" src="https://user-images.githubusercontent.com/90871/146199044-a9b90d92-bec1-4a10-94ce-1fe4ff7d985c.png">

So this can be marked as QA ✅

**Comment by felixarntz** [comment]
### Approval ❌ 

The Idea Hub notice now no longer shows for me in the editor for Idea Hub drafts, with a production build. This is because the `googlesitekit-idea-hub-notice.js` file throws a 404 - it looks like that's happening because it's now no longer present in the manifest file (WordPress tries to load the file without the hash suffix, which confirms that it's missing there).

cc @kuasha420 @tofumatt @wpdarren

**Comment by aaemnnosttv** [comment]
@felixarntz fixed in #4661

**Comment by wpdarren** [comment]
### QA Update: ✅ 

Can confirm that the issue reported is fixed.  When I click on `View draft` link in the Drafts tab, I am redirected to the post editor and the notification now appears and no 404 errors are in the console.

![image](https://user-images.githubusercontent.com/73545194/149541971-8395cdda-6fc2-473c-be25-550d0a1bf710.png)

cc. @felixarntz @aaemnnosttv 👍
